### PR TITLE
Upgrade GPU CI to PyTorch 1.13

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #15555

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: .azure/gpu-tests-pytorch.yml [hunk 1]
```diff
@@ -44,7 +44,7 @@ jobs:
       matrix:
         # TODO: package parametrization
         'PyTorch - stable':
-          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.12-cuda11.6.1"
```
Comment (awaelchli) [inline .azure/gpu-tests-pytorch.yml:47]: how come we had 11.6.1 here while the docker file had 11.3.1?

Comment (carmocca) [reply .azure/gpu-tests-pytorch.yml:47]: Because that's the dockerfile default value. The images are constructed from `.github/workflows/release-docker.yml` which you need to update as I did in https://github.com/Lightning-AI/lightning/pull/15347/files#diff-aa38a2216a456b6851247d357e99f30026306296d43bb3c1f843a6614524ef77R22

**Review by carmocca — COMMENTED** [review]
LGTM. Fine updating this first separately to the rest of CI.

Will approve once CI is green and it's actually using 1.13

File: dockers/base-cuda/Dockerfile [hunk 2]
```diff
@@ -13,12 +13,13 @@
...
 ARG UBUNTU_VERSION=20.04
-ARG CUDA_VERSION=11.3.1
+ARG CUDA_VERSION=11.7
+
 
 FROM nvidia/cuda:${CUDA_VERSION}-devel-ubuntu${UBUNTU_VERSION}
 
 ARG PYTHON_VERSION=3.9
-ARG PYTORCH_VERSION=1.12
+ARG PYTORCH_VERSION=1.13
```
Comment (Borda) [inline dockers/base-cuda/Dockerfile:22]: I think there is some issue with Bagua, https://github.com/Lightning-AI/lightning/pull/15403#issuecomment-1297402599

**Comment by awaelchli** [comment]
How naive of me to think I could just update a few variables from 1.12 to 1.13. I'm just digging my own grave here if I continue.  Maybe this isn't as important as I thought. Sorry for the noise.

**Comment by carmocca** [comment]
I'm hijacking this

**Comment by carmocca** [comment]
CI stuff is ready, now blocked by a failing FSDP test: https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=115069&view=logs&j=3f274fac-2e11-54ca-487e-194c91f3ae9f&t=8e4ceb7c-ceed-5ee0-b6fc-c9023e41cb74&l=2191 cc @awaelchli

**Comment by awaelchli** [comment]
I can take a look too this afternoon!

**Comment by Borda** [comment]
> I can take a look too this afternoon!

as spoke, let's run PT 1.12 with all strategies and then PT 1.13 just for dependencies available at this time as running with 1.13 is important for us for testing the recent FSDP feature...

File: .azure/gpu-tests-pytorch.yml [hunk 3]
```diff
@@ -90,30 +93,34 @@ jobs:
...
         FREEZE_REQUIREMENTS: "1"
-      displayName: 'Install package'
+      displayName: 'Install package & extras'
 
     - bash: |
         set -e
         CUDA_VERSION_BAGUA=$(python -c "print([ver for ver in [116,113,111,102] if $CUDA_VERSION_MM >= ver][0])")
         pip install "bagua-cuda$CUDA_VERSION_BAGUA"
 
+        # TODO: 1.13 wheels are not released, uncomment once they are
```
Comment (carmocca) [inline .azure/gpu-tests-pytorch.yml:107]:
```suggestion
```

**Comment by awaelchli** [comment]
@carmocca @Borda It seems that PT 1.13 is setting some env vars that trigger our leakage assertions. These are the names:
```
KMP_INIT_AT_FORK  
KMP_DUPLICATE_LIB_OK
```
Shall we add them to the allowlist? There seem to be several tests that experience this leakage.

**Comment by carmocca** [comment]
Yes

**Review by carmocca — APPROVED** [review]
LGTM once fsdp tests are fixed

**Comment by awaelchli** [comment]
@carmocca FYI, the reshard_after_forward attribute that caused the attribute error was removed in PyTorch 1.13: https://github.com/pytorch/pytorch/pull/83665
I need to study the the commit in PyTorch to figure out what the alternative is to testing this. For now, I have removed the assertions from the tests in case you want to merge the PR. I can do this investigation as part of #14898

File: dockers/base-cuda/Dockerfile [hunk 4]
```diff
@@ -36,7 +36,12 @@ ENV \
 RUN \
     # TODO: Remove the manual key installation once the base image is updated.
     # https://github.com/NVIDIA/nvidia-docker/issues/1631
-    apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub && \
+    # https://github.com/NVIDIA/nvidia-docker/issues/1631#issuecomment-1264715214
+    wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub && \
+    sudo cp 3bf863cc.pub /etc/apt/keyrings/ && \
+    sudo echo "deb [signed-by=/etc/apt/keyrings/3bf863cc.pub] https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64/ /" /etc/apt/sources.list.d/cuda.list && \
+    sudo apt-get update && \
+    sudo apt-get install cuda && \
```
Comment (Borda) [inline dockers/base-cuda/Dockerfile]: this could be a problem as we are already using docker with cuda

Comment (awaelchli) [reply dockers/base-cuda/Dockerfile]: I have absolutely no idea what I'm doing btw

File: .github/workflows/ci-pytorch-dockers.yml [hunk 5]
```diff
@@ -2,7 +2,7 @@ name: Docker
 
 on:
   push:
-    branches: [master, "release/*"]
+    branches: ["*"]  # FIXME
```
Comment (carmocca) [inline .github/workflows/ci-pytorch-dockers.yml:5]:
```suggestion
    branches: [master, "release/*"]
```

File: .github/workflows/ci-pytorch-dockers.yml [hunk 6]
```diff
@@ -24,7 +24,7 @@ concurrency:
   cancel-in-progress: ${{ ! (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/')) }}
 
 env:
-  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
+  PUSH_TO_HUB: true # FIXME ${{ github.event_name == 'schedule' }}
```
Comment (carmocca) [inline .github/workflows/ci-pytorch-dockers.yml:27]:
```suggestion
  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
```

File: dockers/base-cuda/Dockerfile [hunk 7]
```diff
@@ -132,24 +138,32 @@ RUN \
...
+        CUDA_VERSION_COLOSSALAI=$(python -c "print([ver for ver in [11.3, 11.1] if $CUDA_VERSION_MM_COLOSSALAI >= ver][0])") ; \
+        pip install "colossalai==0.1.10+torch${PYTORCH_VERSION_COLOSSALAI}cu${CUDA_VERSION_COLOSSALAI}" --find-links https://release.colossalai.org ; \
+        python -c "import colossalai; print(colossalai.__version__)" ; \
+    fi
 
 RUN \
     # install rest of strategies
     # remove colossalai from requirements since they are installed separately
     python -c "fname = 'requirements/pytorch/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'colossalai' not in line] ; open(fname, 'w').writelines(lines)" ; \
+    python -c "fname = 'requirements/pytorch/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'horovod' not in line] ; open(fname, 'w').writelines(lines)" ; \
```
Comment (awaelchli) [inline dockers/base-cuda/Dockerfile:166]: In the job above, we were compiling horovod from source (latest version). Here in the strategy.txt file however, the version was set to <0.25.1 which triggered a re-install, without cuda support. That's why we saw the horovod tests failing.

File: .github/checkgroup.yml [hunk 8]
```diff
@@ -128,12 +128,14 @@ subprojects:
...
       - "build-cuda (3.9, 1.11, 11.3.1)"
       - "build-cuda (3.9, 1.12, 11.6.1)"
+      - "build-cuda (3.9, 1.13, 11.7.0)"
       - "build-hpu (1.5.0, 1.11.0)"
       - "build-ipu (3.9, 1.10)"
       - "build-NGC"
       - "build-pl (3.9, 1.10, 11.3.1)"
       - "build-pl (3.9, 1.11, 11.3.1)"
       - "build-pl (3.9, 1.12, 11.6.1)"
+      # TODO: add 1.13
```
Comment (awaelchli) [inline .github/checkgroup.yml:138]: @carmocca what about these todos that you added?
what is the difference between build-cuda and build-pl?

File: .azure/gpu-benchmark.yml [hunk 9]
```diff
@@ -46,19 +46,43 @@ jobs:
 
     - bash: |
         echo "##vso[task.setvariable variable=CUDA_VISIBLE_DEVICES]$(DEVICES)"
-        cuda_ver=$(python -c "import torch ; print(''.join(map(str, torch.version.cuda.split('.')[:2])))")
-        echo "##vso[task.setvariable variable=CUDA_VERSION_MM]$cuda_ver"
-        echo "##vso[task.setvariable variable=TORCH_URL]https://download.pytorch.org/whl/cu${cuda_ver}/torch_stable.html"
+        echo "##vso[task.setvariable variable=TORCH_URL]https://download.pytorch.org/whl/cu116/torch_stable.html"
```
Comment (awaelchli) [inline .azure/gpu-benchmark.yml:49]: The cuda_ver variable here did not have a value. It caused torchvision to be installed with +cu117 but we needed cu116 to match the torch version. Hardcoding it here since I didn't know what else to do

File: .github/checkgroup.yml [hunk 10]
```diff
@@ -128,12 +128,14 @@ subprojects:
       - "build-cuda (3.9, 1.10, 11.3.1)"
       - "build-cuda (3.9, 1.11, 11.3.1)"
       - "build-cuda (3.9, 1.12, 11.6.1)"
+      - "build-cuda (3.9, 1.13, 11.7.0)"
```
Comment (awaelchli) [inline .github/checkgroup.yml:131]:
```suggestion
      - "build-cuda (3.9, 1.13, 11.6.1)"
```

File: .github/checkgroup.yml (hunk 8 above)
Comment (carmocca) [reply .github/checkgroup.yml:138]: The latter is for cpu jobs and should be left for Jirka's PR

File: .azure/gpu-benchmark.yml (hunk 9 above)
Comment (carmocca) [reply .azure/gpu-benchmark.yml:49]: You can copy it from the other workflows. Removing it was probably an accident

Comment (awaelchli) [reply .azure/gpu-benchmark.yml]: that worked!

File: .github/workflows/ci-pytorch-dockers.yml [hunk 11]
```diff
@@ -22,7 +22,7 @@ concurrency:
   cancel-in-progress: ${{ ! (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/')) }}
 
 env:
-  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
+  PUSH_TO_HUB: true
```
Comment (awaelchli) [inline .github/workflows/ci-pytorch-dockers.yml:25]: Before merge:
```suggestion
  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
```

Comment (carmocca) [inline .github/workflows/ci-pytorch-dockers.yml:25]: Reminder to revert this once it's pushed

**Review by carmocca — APPROVED** [review]

File: .github/workflows/release-docker.yml [hunk 12]
```diff
@@ -2,7 +2,7 @@ name: Docker
 
 on:
   push:
-    branches: [master, "release/*"]
+    branches: ["*"]
```
Comment (awaelchli) [inline .github/workflows/release-docker.yml:5]: Before merge:
```suggestion
    branches: [master, "release/*"]
```

File: .azure/gpu-benchmark.yml [hunk 13]
```diff
@@ -49,18 +49,43 @@ jobs:
...
+        which python && which pip
+        python --version
+        pip --version
         pip list
+      displayName: 'Image info & NVIDIA'
+
+    - bash: |
+        python -c "fname = 'requirements/pytorch/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'horovod' not in line] ; open(fname, 'w').writelines(lines)"
+        python -c "fname = 'requirements/pytorch/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'bagua' not in line] ; open(fname, 'w').writelines(lines)"
+        python -c "fname = 'requirements/pytorch/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'colossalai' not in line] ; open(fname, 'w').writelines(lines)"
```
Comment (Borda) [inline .azure/gpu-benchmark.yml:70]:
```suggestion
        python .actions/assistant.py requirements_prune_pkgs --packages [horovod,bagua,colossalai] --req_files [requirements/pytorch/strategies.txt]
```

File: .github/workflows/ci-pytorch-dockers.yml [hunk 14]
```diff
@@ -37,6 +37,7 @@ jobs:
           - {python_version: "3.9", pytorch_version: "1.10", cuda_version: "11.3.1"}
           - {python_version: "3.9", pytorch_version: "1.11", cuda_version: "11.3.1"}
           - {python_version: "3.9", pytorch_version: "1.12", cuda_version: "11.6.1"}
+          # TODO: add 1.13
```
Comment (Borda) [inline .github/workflows/ci-pytorch-dockers.yml:40]:
```suggestion
          - {python_version: "3.9", pytorch_version: "1.13", cuda_version: "11.6.1"}
```
I think this shall be added, right?

File: .azure/gpu-tests-pytorch.yml (hunk 1 above)
Comment (Borda) [reply .azure/gpu-tests-pytorch.yml:44]: but the name for export is based inputs not defaults

File: dockers/base-cuda/Dockerfile [hunk 15]
```diff
@@ -132,24 +138,32 @@ RUN \
...
+        CUDA_VERSION_MM_COLOSSALAI=$(python -c "import torch ; print(''.join(map(str, torch.version.cuda)))") ; \
+        CUDA_VERSION_COLOSSALAI=$(python -c "print([ver for ver in [11.3, 11.1] if $CUDA_VERSION_MM_COLOSSALAI >= ver][0])") ; \
+        pip install "colossalai==0.1.10+torch${PYTORCH_VERSION_COLOSSALAI}cu${CUDA_VERSION_COLOSSALAI}" --find-links https://release.colossalai.org ; \
+        python -c "import colossalai; print(colossalai.__version__)" ; \
+    fi
 
 RUN \
     # install rest of strategies
     # remove colossalai from requirements since they are installed separately
     python -c "fname = 'requirements/pytorch/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'colossalai' not in line] ; open(fname, 'w').writelines(lines)" ; \
```
Comment (Borda) [inline dockers/base-cuda/Dockerfile:165]: if we copy assistant we can simply use:
```
python .actions/assistant.py requirements_prune_pkgs --packages [horovod,colossalai] --req_files [requirements/pytorch/strategies.txt]
```

**Review by Borda — APPROVED** [review]

File: dockers/base-cuda/Dockerfile (hunk 15 above)
Comment (carmocca) [reply dockers/base-cuda/Dockerfile:165]: I suggest you do this change for all workflows in a separate PR. This doesn't need to be done here

Comment (Borda) [reply dockers/base-cuda/Dockerfile:165]: yeah, that is why I did not push it :)
